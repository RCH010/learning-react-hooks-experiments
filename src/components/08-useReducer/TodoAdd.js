import React from 'react'
import PropTypes from 'prop-types'
import { useForm } from '../../hooks/useForm';

const TodoAdd = ({handleAddTodo}) => {
    const [formValues, handleInputChange, reset] = useForm({
        description: ''
    });

    const getNewID = () => {
        return new Date().getTime();
    }
    
    const handleSubmit = (e) => {
        e.preventDefault();
        if (formValues.description.trim().length <= 1) { return; }

        const newTodo = {
            id: getNewID(),
            desc: formValues.description,
            done: false
        };

        handleAddTodo(newTodo)
        reset();
    }

    return (
        <>
            <h2>Agregar</h2>
            <hr/>

            <form onSubmit={handleSubmit}>
                <input 
                    type='text'
                    name='description'
                    placeholder='Hacer...'
                    autoComplete='off'
                    className='form-control'
                    value={formValues.description}
                    onChange={handleInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary btn-block'
                >
                    Agregar
                </button>
            </form>
        </>
    )
}

TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
}

export default TodoAdd
