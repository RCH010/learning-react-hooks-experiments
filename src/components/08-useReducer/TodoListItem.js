import React from 'react'
import PropTypes from 'prop-types'

export const TodoListItem = ({todo, index, handleDelete, handleToggle}) => {
    return (
        <li
            key={index}
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'}`}
                onClick={ () => handleToggle(todo.id)}
            >
                {index+1}. {todo.desc}
            </p>
            <button 
                className='btn btn-danger'
                onClick={() => handleDelete(todo.id)}
            >
                Borrar
            </button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}

export default TodoListItem
