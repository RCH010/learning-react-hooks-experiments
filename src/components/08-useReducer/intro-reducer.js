// este es el initial state
const tareas = [
    {
        id:1,
        todo: 'comprar leche',
        done: false
    }
]
// función para editar el arreglo, reducer
const todoReducer = (state = tareas, action) => {
    if(action?.type === 'add'){
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

//NO modificar directamente el arreglo
//  por ejemplo no usar push

const newTarea = {
    id:2,
    todo: 'comprar café',
    done: false
}
// esta es la acción a hacer, que seria agregar el payload (los nuevos datos)
const action = {
    type: 'add',
    payload: newTarea
}

todos = todoReducer(tareas, action);

console.log(todos);