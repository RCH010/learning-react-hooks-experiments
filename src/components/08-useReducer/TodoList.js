import React from 'react'
import { TodoListItem } from './TodoListItem'
import PropTypes from 'prop-types'

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => (
                    <TodoListItem 
                        todo={todo} 
                        key={todo.id}
                        index={i}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
}

export default TodoList

