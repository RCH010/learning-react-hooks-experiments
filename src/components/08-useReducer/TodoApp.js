import React, { useEffect, useReducer } from 'react'
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer'
import TodoAdd from './TodoAdd';
import './styles.css'


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoReducer, [], init) // puede recibir un init, que seria el edo inicial
    const [todos, dispatch] = useReducer(todoReducer, [], init);
    
    // Con este useEffect, mantenemos los datos actualizados con el local storage,
    //  estos se actualizan cuando cambia el valor de la variable todos
    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])
    
    // cuando se presiona el borrar
    const handleDelete = (todoid) => {
        const action = { type:'delete', payload:todoid };
        dispatch(action);
    }
    // cuando se da click en el parrafo
    const handleToggle = (todoid) => {
        const action = { type:'toggle', payload:todoid };
        dispatch(action);
    }

    const handleAddTodo = (newTodo) => {
        const action = { type: 'add', payload: newTodo };
        dispatch(action);
    }

    return (
        <div>
            <h1>TodoApp ( {todos.length} )</h1>
            <br/>
            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={todos} handleDelete={handleDelete} handleToggle={handleToggle}/>
                </div>
                <div className='col'>
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}
