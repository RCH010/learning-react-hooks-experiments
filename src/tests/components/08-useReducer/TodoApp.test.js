import React from 'react';
import { mount, shallow } from "enzyme"
import { TodoApp } from "../../../components/08-useReducer/TodoApp"
import { demoTodos } from '../../fixitures/demoTodos';
import { act } from '@testing-library/react';

describe('Tests in TodoApp', () => {
    const wrapper = shallow(<TodoApp />);
    Storage.prototype.setItem = jest.fn();

    test('should match TodoApp with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should add one Todo', () => {
        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        })
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 2 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('should delete one Todo', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 1 )');

        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 0 )');
    });
    
})
