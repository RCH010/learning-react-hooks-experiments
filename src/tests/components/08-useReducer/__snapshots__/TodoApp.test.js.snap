// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in TodoApp should match TodoApp with snapshot 1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <br />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
